aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -66,236 +57,344 @@ impl AggregatorID {
     }
 }
 
-/// Tracks values seen by aggregator. In particular, stores information about
-/// the biggest and the smallest deltas seen during execution in the VM. This
-/// information can be used by the executor to check if delta should have
-/// failed. Most importantly, it allows commutativity of adds/subs. Example:
-///
-///
-/// This graph shows how delta of aggregator changed during a single transaction
-/// execution:
-///
-/// ```text
-/// +A ===========================================>
-///            ||
-///          ||||                               +X
-///         |||||  ||||||                    ||||
-///      |||||||||||||||||||||||||          |||||
-/// +0 ===========================================> time
-///                       ||||||
-///                         ||
-///                         ||
-/// -B ===========================================>
-/// ```
-///
-/// Clearly, +X succeeds if +A and -B succeed. Therefore each delta
-/// validation consists of:
-///   1. check +A did not overflow
-///   2. check -A did not drop below zero
-/// Checking +X is irrelevant since +A >= +X.
-///
-/// TODO: while we support tracking of the history, it is not yet fully used on
-/// executor side because we don't know how to throw errors.
-#[derive(Debug)]
-pub struct History {
-    pub max_positive: u128,
-    pub min_negative: u128,
+/// Describes how the `speculative_start_value` in `AggregatorState` was obtained.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum SpeculativeStartValue {
+    // The speculative_start_value is assigned to 0.
+    Unset,
+    // The speculative_start_value was obtained by reading
+    // the last committed value of the aggregator from MVHashmap.
+    LastCommittedValue(u128),
+    // The speculative_start_value was obtained by performing a read
+    // procedure on the aggregator, which involves aggregating deltas
+    // present at the read time
+    AggregatedValue(u128),
 }
 
-impl History {
-    fn new() -> Self {
-        History {
-            max_positive: 0,
-            min_negative: 0,
+impl SpeculativeStartValue {
+    pub fn get_any_value(&self) -> PartialVMResult<u128> {
+        match self {
+            SpeculativeStartValue::Unset => Err(
```

Comment:
The comment says that "speculative start value is assigned to 0", but I feel I do like not returning 0 here better. Should we change the comment maybe?